ouroboros-consensus/src/Ouroboros/Consensus/Util/ThreadRegistry.hs

```diff
@@ -0,0 +1,217 @@
+  { _registered       :: !(TVar m (Map RegisteredID (Registered m)))
+    -- ^ The registered threads and subregistries.
+    --
+    -- The 'RegisteredID' is used to remove a terminated thread or a
+    -- subregistry from this map.
+  , _nextRegisteredID :: !(TVar m RegisteredID)
+    -- ^ The value to use for the next 'RegisteredID' that will be added to
+    -- '_registered'.
+  }
+
+-- | Create a new 'ThreadRegistry' and pass it to the supplied function.
+--
+-- Using the 'ThreadRegistry' you can:
+--
+-- * 'fork': spawn a new thread and get an 'Async' handle for the thread.
+-- * 'forkLinked': same as 'fork' but any exception thrown in the thread will
+--    be thrown to the thread that forked it.
+-- * 'forkNonTerminating': same as 'forkLinked' but meant for non-terminating
+--    actions.
+-- * 'withSubregistry': create subregistries.
+--
+-- When the supplied continuation function returns or throws an exception, all
+-- threads created using the 'ThreadRegistry', as well as all threads created
+-- using subregistries created using the 'ThreadRegistry' will be terminated
+-- using 'uninterruptibleCancel', similar to 'withAsync'.
+withThreadRegistry :: (MonadAsync m, MonadThrow m)
+                   => (ThreadRegistry m -> m a)
+                   -> m a
+withThreadRegistry = bracket (atomically new) cancelAll
+
+-- | Fork a new thread.
+--
+-- You can use the returned 'Async' handle to wait on it and to obtain the
+-- result or exception.
+--
+-- The thread registry will cancel the thread when the thread registry is shut
+-- down.
+fork :: (MonadAsync m, MonadMask m)
+     => ThreadRegistry m
+     -> m a  -- ^ The action to run in the thread
+     -> m (Async m a)
+fork threadRegistry action = mask $ \restore -> do
+    -- We obtain the registeredID in a separate transaction, before the one
+    -- that stores the handle in the map, because we need to know the
+    -- registeredID in the action, which returns the actual handle.
+    registeredID <- atomically $ getNextRegisteredID threadRegistry
+
+    handle <- async $
+      -- When the action terminates successfully or not, we deregister it as
+      -- we will no longer need to cancel it.
+      restore action `finally`
+        atomically (deregister registeredID threadRegistry)
+
+    atomically $ register registeredID (AsyncThread handle) threadRegistry
+
+    return handle
+
+-- | Same as 'fork', but any exception thrown in the forked thread will be
+-- rethrown wrapped in an 'ExceptionInForkedThread' to the thread that called
+-- this function, like 'link'.
+forkLinked :: forall m a. (MonadAsync m, MonadFork m, MonadMask m)
+           => ThreadRegistry m
+           -> m a  -- ^ The action to run in the thread
+           -> m (Async m a)
+forkLinked threadRegistry action = do
+    me <- myThreadId
+    fork threadRegistry $
+      action `catch` \(e :: SomeException) -> do
+        unless (isCancel (Proxy @m) e) $ throwTo me e
+        throwM e
```

Comment:
Would it be feasible to use the existing `link` definition?

Reply:
Judging by the implementation of the real `link`, this solution seems much simpler. I'm open to suggestions.